### PR TITLE
Add https option for istio

### PR DESCRIPTION
**Pull request by michaelcheah**
Closes https://github.com/SeldonIO/ansible-k8s-collection/issues/13

### CODE REVIEW

File: roles/istio/templates/gateway.yaml.j2 [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    app: istio-ingressgateway
+    istio: ingressgateway
+  servers:
+  - hosts:
+    - '*'
+    port:
+      name: http
+      number: 80
+      protocol: HTTP
+{%- if istio_secure %}
```
Comment (RafalSkolasinski) [inline roles/istio/templates/gateway.yaml.j2:17]: What is `-` doing after `{%`?

File: roles/istio/tasks/create-gateway-ssl-secret.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
+#!/bin/bash
```
Comment (RafalSkolasinski) [inline roles/istio/tasks/create-gateway-ssl-secret.sh:1]: This would need to be ansibleized. Worth checking if there is an Ansible modules that will help to generate ssl certificates.

File: roles/istio/defaults/main.yaml [hunk 3]
```diff
@@ -2,3 +2,4 @@
 istio_version: 1.9.0
 istio_verify_install: true
 istio_create_seldon_gateway: true
+istio_secure: false
```
Comment (RafalSkolasinski) [inline roles/istio/defaults/main.yaml:5]: istio_use_ssl? :) Not a strong opinion but I think we were using use_ssl in other places.

File: roles/istio/templates/gateway.yaml.j2 (hunk 1 above)
Comment (michaelcheah) [reply roles/istio/templates/gateway.yaml.j2:17]: It just removes whitespace. It's similar to golang templating

You can check out an example here https://cryptic-cliffs-32040.herokuapp.com/ (used this for testing)

File: roles/istio/tasks/create-gateway-ssl-secret.sh (hunk 2 above)
Comment (RafalSkolasinski) [reply roles/istio/tasks/create-gateway-ssl-secret.sh:1]: https://docs.ansible.com/ansible/latest/collections/community/crypto/index.html

File: roles/istio/tasks/certificates.yaml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    subject_alt_name:
+      - "DNS:istio-ingressgateway" # TODO: put istio-ingressgateway and istio-system into variables? {{service}} {{namespace}}
+      - "DNS:istio-ingressgateway.istio-system"
+      - "DNS:istio-ingressgateway.istio-system.svc"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster.local"
+
+
+# TODO: Original command generates a ca.srl file that is from the CAcreateserial flag.
+#       This is not generated with these steps at the moment but may not be required
```
Comment (michaelcheah) [inline roles/istio/tasks/certificates.yaml:88]: To explain, the original bash scripts generated the x509 cert as follows
```
# Self sign
openssl x509 -req -days 365 -in ${tmpdir}/server.csr -CA ${tmpdir}/ca.crt -CAkey ${tmpdir}/ca.key -CAcreateserial -out ${tmpdir}/server.crt -extfile ${tmpdir}/csr.conf
```
Technically, I think this is not actually "self signed". It's actually using its own CA. The CA itself is self-signed

Beyond that, there is a `CAcreateserial` flag that is passed when generating the x509 cert, which according to the docs `-CAcreateserial            Create serial number file if it does not exist` that I have not been able to incorporate here.

I'm not sure how important it is. https://gist.github.com/Soarez/9688998#signing

**Review by michaelcheah — COMMENTED** [review]
Some thought needs to go into how we expose generating these certificates, or whether we even allow users to generate their certificates via this ansible role. There are several ways for users to provide get their certificates/keys signed, so catering to all users might be very complex. What has been provided here so far is a self-signed CA version, which would probably not be acceptable for production, but would be very useful for quick testing

The way I see it, there's two options moving forward:
~~1. Allow users to only provide the server cert and private key, which will be added as secrets that istio-ingressgateway can access~~
2. Allow users to pass along their CA details, and expose all ansible's `x509_certificate` fields for generating the server certs/keys from there. 

~~I'm inclined to go ahead with 1., as I think it's very out of scope for the istio role to be configurable for users to generate server certs/keys. The current work can be used as follows:~~
```yaml
istio_use_ssl: true
istio_use_self_signed_ca: true
```

~~And then if users want to provide their own server certs and keys:~~
```yaml
istio_server_cert_file: /path/to/server.crt
istio_server_key_file: /path/to/server.key 
```

~~Or users can go ahead and create their own `secret`, if we specify in the production docs to make sure this is stated~~
```bash
kubectl create secret tls ${secret} \
        --key=path/to/server.key \
        --cert=path/to/server.crt \
        --dry-run=client -o yaml |
    kubectl -n ${namespace} apply -f -
```

EDIT: On further inspection, I realise that the `server.crt` requires a few details, including information like `-subj "/CN=${host}"` and the `subjectAltName` list. So I realise there's actually only one option, i.e, allow users to pass along their CA cert and key for ansible to generate the server key and certificate, which is then stored as a secret.

So our configs would be
```yaml
istio_use_ssl: true # defaults to false

# either istio_ca_self_sign is true, or both istio_ca_cert_file and istio_ca_key_file is filled
istio_ca_self_sign: true 
istio_ca_cert_vars: /path/to/x509_cert_config.yaml
```

Where `x509_cert_config.yaml`:
```yaml
# We configure these two paths for the user
# path: /etc/ssl/crt/ansible.com.crt
# csr_path: /etc/ssl/csr/ansible.com.csr

# users configure any of the below depending on how they need to access their CA
ownca_path: /etc/ssl/crt/ansible_CA.crt
ownca_privatekey_path: /etc/ssl/private/ansible_CA.pem
provider: ownca
entrust_requester_name: Jo Doe
entrust_requester_email: jdoe@ansible.com
entrust_requester_phone: 555-555-5555
entrust_cert_type: STANDARD_SSL
entrust_api_user: apiusername
entrust_api_key: a^lv*32!cd9LnT
entrust_api_client_cert_path: /etc/ssl/entrust/ecs-client.crt
entrust_api_client_cert_key_path: /etc/ssl/entrust/ecs-key.crt
entrust_api_specification_path: /etc/ssl/entrust/api-docs/cms-api-2.1.0.yaml
...
```
### tl;dr
In other words

Our job: 
  * Create the server private key
  * Generate the CSR based on configs that we require e.g. `subjectAltName` -> `server.csr`
  * Store the generated `server.crt` and `server.key` as kubernetes secrets that `istio-ingressgateway` can access

User's job:
  * Configure how CA cert is used to sign the certificate by following https://docs.ansible.com/ansible/latest/collections/community/crypto/x509_certificate_module.html#examples

File: roles/istio/tasks/certificates.yaml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+      - digitalSignature
+      - keyEncipherment
+    extended_key_usage:
+      - serverAuth
+    subject_alt_name:
+      - "DNS:istio-ingressgateway"
+      - "DNS:istio-ingressgateway.istio-system"
+      - "DNS:istio-ingressgateway.istio-system.svc"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster.local"
```
[the remark this answers is not in the document]
Comment (michaelcheah) [reply roles/istio/tasks/certificates.yaml:108]: I realise these values in the `x509_certificate` generation are used for **assertions**, which might be useful to keep

File: roles/istio/tasks/certificates.yaml [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+#    wait_timeout: 360
+#  register: istio_web_service
+#  until: istio_web_service.resources[0].status.loadBalancer.ingress[0].ip != ""
+#  retries: 5
+#  delay: 2
+
+
+- name: Set fact about Istio Ingress Gateway External IP
+  set_fact:
+    istio_external_ip: "35.123.14.14141" #"{{ istio_web_service.resources[0].status.loadBalancer.ingress[0].ip }}"
```
Comment (michaelcheah) [inline roles/istio/tasks/certificates.yaml:21]: I forgot to undo this bit (was testing creating certs without istio) while I was doing the real test.

It looks like this info isn't strictly required. Need to investigate why

File: roles/istio/tasks/certificates.yaml [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      - "DNS:istio-ingressgateway.istio-system"
+      - "DNS:istio-ingressgateway.istio-system.svc"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster"
+      - "DNS:istio-ingressgateway.istio-system.svc.cluster.local"
+
+
+# TODO: Original command generates a ca.srl file that is from the CAcreateserial flag.
+#       This is not generated with these steps at the moment but may not be required
+- name: Generate an OpenSSL certificate from the server CSR signed with our CA certificate
+  community.crypto.x509_certificate:
```
Comment (michaelcheah) [inline roles/istio/tasks/certificates.yaml:90]: Find out how to expire certificate after a year (365 days)

File: roles/istio/tasks/certificates.yaml (hunk 6 above)
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:21]: That looks worrying. In other roles I have a logic that reads `istio_external_ip` automatically as that information is sometimes required.

File: roles/istio/tasks/certificates.yaml [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+################# Start Store Server Cert as Secret #################
+
+- name: Register Server CRT
+  shell: "cat {{ tmp_cert_dir.path }}/server.crt | base64"
+  register: enc_server_crt
+
+
+- name: Register Server Key
+  shell: "cat {{ tmp_cert_dir.path }}/server.key | base64"
+  register: enc_server_key
```
[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:121]: I believe I have a logic somewhere that uses a `lookup` to pipe into base64

[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:121]: Example
```yaml
- name: Create Realm Secret
  community.kubernetes.k8s:
    state: present
    namespace: "{{ keycloak_namespace }}"
    definition:
      apiVersion: v1
      kind: Secret
      metadata:
        name: realm-secret
      type: Opaque
      data:
        realm-export.json: "{{ lookup('file', 'realm-export.json') | b64encode }}"
```

[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:121]: That should work :)

File: roles/istio/tasks/certificates.yaml [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    namespace: istio-system
+    definition:
+      apiVersion: v1
+      kind: Secret
+      metadata:
+        name: istio-ingressgateway-certs
+      type: kubernetes.io/tls
+      data:
+        tls.crt: "{{ enc_server_crt.stdout }}"
+        tls.key: "{{ enc_server_key.stdout }}"
```
[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:136]: Template will be better indeed

File: roles/istio/tasks/certificates.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (michaelcheah) [reply roles/istio/tasks/certificates.yaml:111]: Note that 
> The assertonly provider is intended for use cases where one is only interested in checking properties of a supplied certificate. Please note that this provider has been deprecated in Ansible 2.9 and will be removed in community.crypto 2.0.0. See the examples on how to emulate assertonly usage with community.crypto.x509_certificate_info, community.crypto.openssl_csr_info, community.crypto.openssl_privatekey_info and ansible.builtin.assert. This also allows more flexible checks than the ones offered by the assertonly provider.

Eventually, I will want to follow the example here:
```yaml
# The following example shows one assertonly usage using all existing options for
# assertonly, and shows how to emulate the behavior with the x509_certificate_info,
# openssl_csr_info, openssl_privatekey_info and assert modules:
- name: Usage of assertonly with all existing options
  community.crypto.x509_certificate:
    provider: assertonly
    path: /etc/ssl/crt/ansible.com.crt
    csr_path: /etc/ssl/csr/ansible.com.csr
    privatekey_path: /etc/ssl/csr/ansible.com.key
    signature_algorithms:
      - sha256WithRSAEncryption
      - sha512WithRSAEncryption
    subject:
      commonName: ansible.com
    subject_strict: yes
    issuer:
      commonName: ansible.com
    issuer_strict: yes
    has_expired: no
    version: 3
    key_usage:
      - Data Encipherment
    key_usage_strict: yes
    extended_key_usage:
      - DVCS
    extended_key_usage_strict: yes
    subject_alt_name:
      - dns:ansible.com
    subject_alt_name_strict: yes
    not_before: 20190331202428Z
    not_after: 20190413202428Z
    valid_at: "+1d10h"
    invalid_at: 20200331202428Z
    valid_in: 10  # in ten seconds

- name: Get certificate information
  community.crypto.x509_certificate_info:
    path: /etc/ssl/crt/ansible.com.crt
    # for valid_at, invalid_at and valid_in
    valid_at:
      one_day_ten_hours: "+1d10h"
      fixed_timestamp: 20200331202428Z
      ten_seconds: "+10"
  register: result

- name: Get CSR information
  community.crypto.openssl_csr_info:
    # Verifies that the CSR signature is valid; module will fail if not
    path: /etc/ssl/csr/ansible.com.csr
  register: result_csr

- name: Get private key information
  community.crypto.openssl_privatekey_info:
    path: /etc/ssl/csr/ansible.com.key
  register: result_privatekey

- assert:
    that:
      # When private key is specified for assertonly, this will be checked:
      - result.public_key == result_privatekey.public_key
      # When CSR is specified for assertonly, this will be checked:
      - result.public_key == result_csr.public_key
      - result.subject_ordered == result_csr.subject_ordered
      - result.extensions_by_oid == result_csr.extensions_by_oid
      # signature_algorithms check
      - "result.signature_algorithm == 'sha256WithRSAEncryption' or result.signature_algorithm == 'sha512WithRSAEncryption'"
      # subject and subject_strict
      - "result.subject.commonName == 'ansible.com'"
      - "result.subject | length == 1"  # the number must be the number of entries you check for
      # issuer and issuer_strict
      - "result.issuer.commonName == 'ansible.com'"
      - "result.issuer | length == 1"  # the number must be the number of entries you check for
      # has_expired
      - not result.expired
      # version
      - result.version == 3
      # key_usage and key_usage_strict
      - "'Data Encipherment' in result.key_usage"
      - "result.key_usage | length == 1"  # the number must be the number of entries you check for
      # extended_key_usage and extended_key_usage_strict
      - "'DVCS' in result.extended_key_usage"
      - "result.extended_key_usage | length == 1"  # the number must be the number of entries you check for
      # subject_alt_name and subject_alt_name_strict
      - "'dns:ansible.com' in result.subject_alt_name"
      - "result.subject_alt_name | length == 1"  # the number must be the number of entries you check for
      # not_before and not_after
      - "result.not_before == '20190331202428Z'"
      - "result.not_after == '20190413202428Z'"
      # valid_at, invalid_at and valid_in
      - "result.valid_at.one_day_ten_hours"  # for valid_at
      - "not result.valid_at.fixed_timestamp"  # for invalid_at
      - "result.valid_at.ten_seconds"  # for valid_in
```

File: roles/istio/tasks/certificates.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (michaelcheah) [reply roles/istio/tasks/certificates.yaml]: 02773f5

File: roles/istio/tasks/certificates.yaml (hunk 7 above)
Comment (michaelcheah) [reply roles/istio/tasks/certificates.yaml:92]: 67d58b1

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]

File: roles/istio/tasks/certificates.yaml [hunk 10]
```diff
@@ -0,0 +1,137 @@
+---
+
+################### Start Get Istio External IP ###################
```
Comment (RafalSkolasinski) [inline roles/istio/tasks/certificates.yaml:3]: Do we need the comments for each section? These should be self explanatory. 
If you think comments are required here we could have a paragraph / bullet points at the beginning of the file maybe?

File: roles/istio/tasks/certificates.yaml [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  until: >
+    'ingress' in istio_web_service.resources[0].status.loadBalancer and
+    istio_web_service.resources[0].status.loadBalancer.ingress[0].ip != ""
+  retries: 20
+  delay: 5
+
+
+- name: Set fact about Istio Ingress Gateway External IP
+  set_fact:
+    istio_external_ip: "{{ istio_web_service.resources[0].status.loadBalancer.ingress[0].ip }}"
```
Comment (RafalSkolasinski) [inline roles/istio/tasks/certificates.yaml:23]: We're using this operation quite often. Let's open issue to explore how to add it as part of the collection's built-in so we can benefit from code re-usability

File: roles/istio/tasks/certificates.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:111]: Nice, good to know but I don't think we need to this extend for now. Basic self-signed certs are good for most cases

File: roles/istio/defaults/main.yaml [hunk 12]
```diff
@@ -2,3 +2,4 @@
 istio_version: 1.9.0
 istio_verify_install: true
 istio_create_seldon_gateway: true
+istio_use_ssl: false
```
Comment (RafalSkolasinski) [inline roles/istio/defaults/main.yaml:5]: I thought you still wanted to add option to not deploy self signed certs? 
```yaml
istio_use_ssl: true
istio_use_self_signed_ca: {{ istio_use_ssl }}
```
?

File: roles/istio/tasks/certificates.yaml [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+      - "CA:TRUE"
+    basic_constraints_critical: true
+
+
+- name: Generate CA Self Signed Certificate
+  community.crypto.x509_certificate:
+    path: "{{ tmp_cert_dir.path }}/ca.crt"
+    privatekey_path: "{{ tmp_cert_dir.path }}/ca.key"
+    csr_path: "{{ tmp_cert_dir.path }}/ca.csr"
+    provider: selfsigned
```
[the remark this answers is not in the document]
Comment (RafalSkolasinski) [reply roles/istio/tasks/certificates.yaml:57]: > This probably needs to be wrapped in a block so that it is optionally allowed, but the user should be able to pass in their own CA. 

Let's open a follow up issue to allow users to provide their own certs and we will pick it up if the need  arise

File: roles/istio/defaults/main.yaml (hunk 12 above)
Comment (RafalSkolasinski) [reply roles/istio/defaults/main.yaml:5]: Turns out this being optional is quite important.

**Review by RafalSkolasinski — CHANGES_REQUESTED** [review]
Nice work! Just have few tiny comments

File: roles/istio/tasks/main.yaml [hunk 14]
```diff
@@ -41,19 +41,6 @@
...
-    - name: Generate certificates
-      include: certificates.yaml
-  always:
-    - name: Remove temporary certificates directory
-      ansible.builtin.file:
-        path: "{{ tmp_cert_dir.path }}"
-        state: absent
-      when: tmp_cert_dir.path is defined
+  include: certificates.yaml
   when: >
```
Comment (RafalSkolasinski) [inline roles/istio/tasks/main.yaml:45]: one line? :)
